### PR TITLE
[WIP] Expose GPIO 

### PR DESCRIPTION
**Pull request by livanov93**

### CODE REVIEW

File: ur_controllers/src/gpio_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+    set_speed_slider_srv_ = get_node()->create_service<ur_msgs::srv::SetSpeedSliderFraction>(
+        "set_io", std::bind(&GPIOController::setSpeedSlider, this, std::placeholders::_1, std::placeholders::_2));
+  }
+  catch (...)
+  {
+    return LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  return LifecycleNodeInterface::on_configure(previous_state);
```
Comment (destogl) [inline ur_controllers/src/gpio_controller.cpp:152]: I think you should simply return:

```suggestion
  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
```

File: ur_controllers/src/gpio_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  {
+    safety_mode_msg_.mode = safety_mode;
+    safety_mode_pub_->publish(safety_mode_msg_);
+  }
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ur_controllers::GPIOController::on_activate(const rclcpp_lifecycle::State& previous_state)
+{
+  return LifecycleNodeInterface::on_activate(previous_state);
```
Comment (destogl) [inline ur_controllers/src/gpio_controller.cpp:226]: here also only:

```suggestion
  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
```

File: ur_controllers/src/gpio_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ur_controllers::GPIOController::on_deactivate(const rclcpp_lifecycle::State& previous_state)
+{
+  return LifecycleNodeInterface::on_deactivate(previous_state);
```
Comment (destogl) [inline ur_controllers/src/gpio_controller.cpp:232]:
```suggestion
  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
```

**Review by destogl — COMMENTED** [review]
This looks good. Just a few styling comments and readability improvements.

**Comment by destogl** [comment]
@MGBla could you test this on a real hardware?

File: ur_controllers/controller_plugins.xml [hunk 4]
```diff
@@ -5,13 +5,18 @@
...
+    <description>
+      This controller publishes the readings of all available speed scaling factors.
+    </description>
   </class>
   <class name="ur_controllers/ScaledJointTrajectoryController" type="ur_controllers::ScaledJointTrajectoryController" base_class_type="controller_interface::ControllerInterface">
-  <description>
-    Scaled position-based joint trajectory controller
-  </description>
+    <description>
+      Scaled position-based joint trajectory controller
```
Comment (destogl) [inline ur_controllers/controller_plugins.xml:14]:
```suggestion
      Scaled position-based joint trajectory controller.
```

**Comment by livanov93** [comment]
> @MGBla could you test this on a real hardware?

That would be great. Because I am having really bad time testing it on the docker simulation.

**Comment by AndyZe** [comment]
Replaced by #66

**Pull request closed without merging**